File: Sources/MooreRoutines/RoutineEditorBuffer.swift
// contractId: SC-routines @1.0.0
// In-progress routine editor state (§2a). Pure struct: holds the working name + the
// working set list; `applyChanges()` writes back through RoutineDAO in one transaction
// (reorder / add / remove / change planned values — V2). No SwiftUI.

import Foundation

public struct RoutineEditorBuffer: Equatable, Sendable {
    /// The routine being edited. Nil = creating a new routine (applyChanges → create).
    public private(set) var routineId: String?
    public var name: String
    /// Working set list, in display order (index = sortOrder). May be empty (§2a draft
    /// is legal; such a routine has `startEnabled = false` per BR-001).
    public private(set) var drafts: [EditableSetDraft]

    /// Load an existing routine + its live sets into the buffer.
    public init(routine: Routine, sets: [PlannedSet]) {
        self.routineId = routine.id
        self.name = routine.name
        self.drafts = sets.map {
            EditableSetDraft(
                id: $0.id,
                exerciseId: $0.exerciseId,
                plannedWeight: $0.plannedWeight,
                plannedReps: $0.plannedReps,
                plannedDuration: $0.plannedDuration,
                setClass: $0.setClass ?? .work
            )
        }
    }

    /// Start a new empty routine buffer.
    public init(name: String = "") {
        self.routineId = nil
        self.name = name
        self.drafts = []
    }

    // MARK: - Mutations (editor gestures)

    /// Add an exercise (chosen via the SC-exercises picker) as one new set row at the
    /// end. The exercise's id is supplied by the picker's callback.
    public mutating func addExercise(_ exerciseId: String) {
        drafts.append(EditableSetDraft(id: Self.newId(), exerciseId: exerciseId))
    }

    /// Append another set row for an exercise already in the buffer (defaults copied
    /// from that exercise's last row, matching #7's add-set affordance).
    public mutating func addSet(forExercise exerciseId: String) {
        let template = drafts.last { $0.exerciseId == exerciseId }
        var row = EditableSetDraft(id: Self.newId(), exerciseId: exerciseId)
        if let template {
            row.plannedWeight = template.plannedWeight
            row.plannedReps = template.plannedReps
            row.plannedDuration = template.plannedDuration
            row.setClass = template.setClass
        }
        drafts.append(row)
    }

    /// Update a single row's planned values (weight / reps / duration / class).
    public mutating func updateSet(id: String, _ mutate: (inout EditableSetDraft) -> Void) {
        guard let i = drafts.firstIndex(where: { $0.id == id }) else { return }
        mutate(&drafts[i])
    }

    /// Remove a set row.
    public mutating func removeSet(id: String) {
        drafts.removeAll { $0.id == id }
    }

    /// Move a set row (reorder). Indices are into the current `drafts` order.
    public mutating func moveSet(from source: Int, to destination: Int) {
        guard drafts.indices.contains(source), drafts.indices.contains(destination), source != destination else { return }
        let item = drafts.remove(at: source)
        drafts.insert(item, at: destination)
    }

    // MARK: - Persist

    /// Write the buffer back. New routine → `RoutineDAO.create`; existing → `update`
    /// (reorder / add / remove / change planned values in one transaction). Returns the
    /// persisted routine.
    @discardableResult
    public func applyChanges(using dao: RoutineDAO, folderId: String? = nil) throws -> Routine {
        if let routineId {
            return try dao.update(id: routineId, name: name, setDrafts: drafts)
        } else {
            return try dao.create(name: name, folderId: folderId, exerciseList: drafts)
        }
    }

    private static func newId() -> String {
        UUID().uuidString.lowercased()
    }
}

File: android/core/src/main/kotlin/com/moore/routines/RoutineEditorBuffer.kt
// contractId: SC-routines @1.0.0
// In-progress routine editor state (§2a). Pure value type: holds the working name
// + the working set list; applyChanges() writes back through RoutineDAO in one
// transaction (reorder / add / remove / change planned values — V2). No UI.
// Mechanical Kotlin port of Sources/MooreRoutines/RoutineEditorBuffer.swift.
package com.moore.routines

import com.moore.foundation.SetClass
import java.util.UUID

class RoutineEditorBuffer private constructor(
    /// The routine being edited. Null = creating a new routine (applyChanges → create).
    var routineId: String?,
    var name: String,
    /// Working set list, in display order (index = sortOrder). May be empty (§2a
    /// draft is legal; such a routine has startEnabled = false per BR-001).
    drafts: List<EditableSetDraft>,
) {
    private val _drafts = drafts.toMutableList()
    val drafts: List<EditableSetDraft> get() = _drafts

    /// Load an existing routine + its live sets into the buffer.
    constructor(routine: Routine, sets: List<PlannedSet>) : this(
        routineId = routine.id,
        name = routine.name,
        drafts = sets.map {
            EditableSetDraft(
                id = it.id,
                exerciseId = it.exerciseId,
                plannedWeight = it.plannedWeight,
                plannedReps = it.plannedReps,
                plannedDuration = it.plannedDuration,
                setClass = it.setClass ?: SetClass.WORK,
            )
        },
    )

    /// Start a new empty routine buffer.
    constructor(name: String = "") : this(routineId = null, name = name, drafts = emptyList())

    // MARK: - Mutations (editor gestures)

    /// Add an exercise (chosen via the SC-exercises picker) as one new set row
    /// at the end.
    fun addExercise(exerciseId: String) {
        _drafts.add(EditableSetDraft(id = newId(), exerciseId = exerciseId))
    }

    /// Append another set row for an exercise already in the buffer (defaults
    /// copied from that exercise's last row, matching #7's add-set affordance).
    /// (Swift label `addSet(forExercise:)`.)
    fun addSetForExercise(exerciseId: String) {
        val template = _drafts.lastOrNull { it.exerciseId == exerciseId }
        val row = EditableSetDraft(id = newId(), exerciseId = exerciseId)
        if (template != null) {
            row.plannedWeight = template.plannedWeight
            row.plannedReps = template.plannedReps
            row.plannedDuration = template.plannedDuration
            row.setClass = template.setClass
        }
        _drafts.add(row)
    }

    /// Update a single row's planned values (weight / reps / duration / class).
    fun updateSet(id: String, mutate: (EditableSetDraft) -> Unit) {
        val i = _drafts.indexOfFirst { it.id == id }
        if (i < 0) return
        mutate(_drafts[i])
    }

    /// Remove a set row.
    fun removeSet(id: String) {
        _drafts.removeAll { it.id == id }
    }

    /// Move a set row (reorder). Indices are into the current drafts order.
    /// (Swift labels `moveSet(from:to:)`.)
    fun moveSet(source: Int, destination: Int) {
        if (source !in _drafts.indices || destination !in _drafts.indices || source == destination) return
        val item = _drafts.removeAt(source)
        _drafts.add(destination, item)
    }

    // MARK: - Persist

    /// Write the buffer back. New routine → RoutineDAO.create; existing → update
    /// (reorder / add / remove / change planned values in one transaction).
    /// Returns the persisted routine.
    fun applyChanges(dao: RoutineDAO, folderId: String? = null): Routine {
        val routineId = this.routineId
        return if (routineId != null) {
            dao.update(id = routineId, name = name, setDrafts = _drafts.toList())
        } else {
            dao.create(name = name, folderId = folderId, exerciseList = _drafts.toList())
        }
    }

    private fun newId(): String = UUID.randomUUID().toString().lowercase()
}
